google_compute_engine/accounts/tests/oslogin_utils_test.py

```diff
@@ -318,36 +249,79 @@ def testUpdateOsLoginUpdateCache(self, mock_time):
-    ]
-    self.assertEqual(mocks.mock_calls, expected_calls)
-    self.assertEqual(return_value, 0)
-
-  def testUpdateOsLoginRedundantDeactivate(self):
-    mocks = mock.Mock()
-    mocks.attach_mock(self.mock_oslogin, 'oslogin')
-    self.mock_oslogin._RunOsLoginControl.return_value = 0
-    self.mock_oslogin._GetStatus.return_value = False
-
-    return_value = oslogin_utils.OsLoginUtils.UpdateOsLogin(
-        self.mock_oslogin, False, two_factor_desired=True)
-    expected_calls = [
-        mock.call.oslogin._GetStatus(two_factor=False),
-        mock.call.oslogin._GetStatus(two_factor=True),
-    ]
-    self.assertEqual(mocks.mock_calls, expected_calls)
-    self.assertEqual(return_value, 0)
+  def testUpdateOsLogin(self, mock_time):
+
+    def _AssertNoUpdate():
+      expected_calls = [
+          mock.call.oslogin._GetStatus(two_factor=False),
+          mock.call.oslogin._GetStatus(two_factor=True),
+      ]
+      self.assertEqual(mocks.mock_calls, expected_calls)
+      self.assertEqual(return_value, 0)
+
+    def _AssertActivated(two_factor=False):
+      params = ['activate', '--twofactor'] if two_factor else ['activate']
+      expected_calls = [
+          mock.call.oslogin._GetStatus(two_factor=False),
+          mock.call.oslogin._GetStatus(two_factor=True),
+          mock.call.logger.info(mock.ANY),
+          mock.call.oslogin._RunOsLoginControl(params),
+          mock.call.oslogin._RunOsLoginNssCache(),
+      ]
+      self.assertEqual(mocks.mock_calls, expected_calls)
+
+    def _AssertDeactivated():
+      expected_calls = [
+          mock.call.oslogin._GetStatus(two_factor=False),
+          mock.call.oslogin._GetStatus(two_factor=True),
+          mock.call.logger.info(mock.ANY),
+          mock.call.oslogin._RunOsLoginControl(['deactivate']),
+          mock.call.oslogin._RemoveOsLoginNssCache(),
+      ]
+      self.assertEqual(mocks.mock_calls, expected_calls)
+
+    def _AssertReactivated():
+      expected_calls = [
+          mock.call.oslogin._GetStatus(two_factor=False),
+          mock.call.oslogin._GetStatus(two_factor=True),
+          mock.call.logger.info(mock.ANY),
+          mock.call.oslogin._RunOsLoginControl(['deactivate']),
+          mock.call.oslogin._RunOsLoginControl(['activate']),
+      ]
+      self.assertEqual(mocks.mock_calls, expected_calls)
+
+    parameters = list(itertools.product([False, True], repeat=4))
+    for (oslogin, two_factor, oslogin_config, two_factor_config) in parameters:
+      mocks = mock.Mock()
+      mocks.attach_mock(self.mock_logger, 'logger')
+      mocks.attach_mock(self.mock_oslogin, 'oslogin')
+      self.mock_oslogin._RunOsLoginControl.return_value = 0
+      self.mock_oslogin._GetStatus.side_effect = [
+          oslogin_config, two_factor_config]
+      mock_time.return_value = 6 * 60 * 60
+      return_value = oslogin_utils.OsLoginUtils.UpdateOsLogin(
+          self.mock_oslogin, oslogin, two_factor_desired=two_factor)
+
+      if oslogin_config:
```

Comment:
i dont think i would be able to maintain this. would it be a lot of work to use a more traditional testing structure with explicitly called tests?